### PR TITLE
hal: Add HAL for the RADIO peripheral

### PR DESCRIPTION
**Pull request by kl-cruz**
Add the RADIO HAL for the nRF SoCs. Supported SoCs:
 - nRF51 series
 - nRF52810
 - nRF52832
 - nRF52840

### CODE REVIEW

File: hal/nrf_radio.h [hunk 1]
```diff
@@ -0,0 +1,1407 @@
...
+        freq = 2400;
+    }
+    freq += NRF_RADIO->FREQUENCY & RADIO_FREQUENCY_FREQUENCY_Msk;
+
+    return freq;
+}
+
+__STATIC_INLINE void nrf_radio_tx_power_set(nrf_radio_txpower_t tx_power)
+{
+    NRF_RADIO->TXPOWER = (tx_power << RADIO_TXPOWER_TXPOWER_Pos);
```
Comment (anangl) [inline hal/nrf_radio.h]: `(uint32_t)tx_power`

File: hal/nrf_radio.h [hunk 2]
```diff
@@ -0,0 +1,1407 @@
...
+                      | ((uint32_t)p_config->plen << RADIO_PCNF0_PLEN_Pos)
+#endif
+#if defined(RADIO_PCNF0_CRCINC_Msk)
+                      | (p_config->crcinc ? (RADIO_PCNF0_CRCINC_Include << RADIO_PCNF0_CRCINC_Pos) :
+                                            (RADIO_PCNF0_CRCINC_Exclude << RADIO_PCNF0_CRCINC_Pos) )
+#endif
+#if defined(RADIO_PCNF0_TERMLEN_Msk)
+                      | ((uint32_t)p_config->termlen << RADIO_PCNF0_TERMLEN_Pos)
+#endif
+);
```
Comment (anangl) [inline hal/nrf_radio.h]: This should be indented.

File: hal/nrf_radio.h [hunk 3]
```diff
@@ -0,0 +1,1407 @@
...
+}
+
+__STATIC_INLINE uint32_t nrf_radio_bcc_get(void)
+{
+    return NRF_RADIO->BCC;
+}
+
+__STATIC_INLINE void nrf_radio_dab_set(uint32_t dab_value, uint8_t segment)
+{
+    NRF_RADIO->DAB[segment] = dab_value;
```
Comment (anangl) [inline hal/nrf_radio.h]: `NRFX_ASSERT(segment < 8);`

And add similar check in the consecutive three functions.

File: hal/nrf_radio.h [hunk 4]
```diff
@@ -0,0 +1,1407 @@
...
+
+__STATIC_INLINE bool nrf_radio_modecnf0_ramp_up_default_get(void)
+{
+    return ((NRF_RADIO->MODECNF0 & RADIO_MODECNF0_RU_Msk) >> RADIO_MODECNF0_RU_Pos) ==
+            RADIO_MODECNF0_RU_Default;
+}
+
+__STATIC_INLINE uint8_t nrf_radio_modecnf0_dtx_get(void)
+{
+    return ((NRF_RADIO->MODECNF0 & RADIO_MODECNF0_DTX_Msk) >> RADIO_MODECNF0_DTX_Pos);
```
Comment (anangl) [inline hal/nrf_radio.h]: Typecast to `uint8_t` to avoid warnings.

File: hal/nrf_radio.h [hunk 5]
```diff
@@ -0,0 +1,1407 @@
...
+
+#if defined(RADIO_SFD_SFD_Msk)
+__STATIC_INLINE void nrf_radio_sfd_set(uint8_t sfd)
+{
+    NRF_RADIO->SFD = ((uint32_t)sfd) << RADIO_SFD_SFD_Pos;
+}
+
+__STATIC_INLINE uint8_t nrf_radio_sfd_get(void)
+{
+    return ((NRF_RADIO->SFD & RADIO_SFD_SFD_Msk) >> RADIO_SFD_SFD_Pos);
```
Comment (anangl) [inline hal/nrf_radio.h]: Typecast to `uint8_t` to avoid warnings.

File: hal/nrf_radio.h [hunk 6]
```diff
@@ -0,0 +1,1407 @@
...
+    NRF_RADIO_TXPOWER_Pos2dBm  = RADIO_TXPOWER_TXPOWER_Pos2dBm,  /**< 2 dBm. */
+#endif
+    NRF_RADIO_TXPOWER_0dBm     = RADIO_TXPOWER_TXPOWER_0dBm,     /**< 0 dBm. */
+    NRF_RADIO_TXPOWER_Neg4dBm  = RADIO_TXPOWER_TXPOWER_Neg4dBm,  /**< -4 dBm. */
+    NRF_RADIO_TXPOWER_Neg8dBm  = RADIO_TXPOWER_TXPOWER_Neg8dBm,  /**< -8 dBm. */
+    NRF_RADIO_TXPOWER_Neg12dBm = RADIO_TXPOWER_TXPOWER_Neg12dBm, /**< -12 dBm. */
+    NRF_RADIO_TXPOWER_Neg16dBm = RADIO_TXPOWER_TXPOWER_Neg16dBm, /**< -16 dBm. */
+    NRF_RADIO_TXPOWER_Neg20dBm = RADIO_TXPOWER_TXPOWER_Neg20dBm, /**< -20 dBm. */
+    NRF_RADIO_TXPOWER_Neg30dBm = RADIO_TXPOWER_TXPOWER_Neg30dBm, /**< -30 dBm. */
+    NRF_RADIO_TXPOWER_Neg40dBm = RADIO_TXPOWER_TXPOWER_Neg40dBm, /**< -40 dBm. */
```
[the remark this answers is not in the document]
Comment (kl-cruz) [reply hal/nrf_radio.h]: Well, it is not in Product Specification but is define in https://github.com/NordicSemiconductor/nrfx/blob/master/mdk/nrf51_deprecated.h#L89
However I'll remove it

**Comment by kl-cruz** [comment]
I introduced your suggestions. Now should be better

File: hal/nrf_radio.h (hunk 2 above)
Comment (kl-cruz) [reply hal/nrf_radio.h]: Aligned

File: hal/nrf_radio.h [hunk 7]
```diff
@@ -0,0 +1,1421 @@
...
+
+__STATIC_INLINE uint32_t nrf_radio_ifs_get(void)
+{
+    return NRF_RADIO->TIFS;
+}
+
+__STATIC_INLINE uint8_t nrf_radio_rssi_sample_get(void)
+{
+    return (uint8_t)(((NRF_RADIO->RSSISAMPLE) & RADIO_RSSISAMPLE_RSSISAMPLE_Msk) >>
+        RADIO_RSSISAMPLE_RSSISAMPLE_Pos);
```
Comment (anangl) [inline hal/nrf_radio.h]: This line should be indented more (to keep consistency).

File: hal/nrf_radio.h [hunk 8]
```diff
@@ -0,0 +1,1421 @@
...
+    NRF_RADIO_STATE_TXDISABLE = RADIO_STATE_STATE_TxDisable, /**< The radio is disabling the transmitter. */
+} nrf_radio_state_t;
+
+/**
+ * @brief Types of RADIO TX power.
+ */
+typedef enum
+{
+#if defined(RADIO_TXPOWER_TXPOWER_Pos8dBm) || defined(__NRFX_DOXYGEN__)
+    NRF_RADIO_TXPOWER_Pos8dBm  = RADIO_TXPOWER_TXPOWER_Pos8dBm,  /**< 8 dBm. */
```
Comment (anangl) [inline hal/nrf_radio.h]: Should't we keep the convention of using only capital letters in introduced names of enumerators?

Comment (kl-cruz) [reply hal/nrf_radio.h]: Well, it will look strange, but surely it's gonna be cosistent with others. I'll push changes to review

**Comment by kl-cruz** [comment]
@anangl please look at changes I've made

**Comment by kl-cruz** [comment]
I introduced some simplifaction of CRC addr incl and tx_power (changed to txpower) to keep consistent with others

**Pull request merged**